src/door.rs

```diff
@@ -0,0 +1,291 @@
+use super::cfg::{Config, Door, User};
+use super::dos::Templates;
+use super::{LaunchArgs, SysopCmdArgs};
+
+use chrono::Local;
+use fs4::FileExt;
+use serde::Serialize;
+use std::env;
+use std::fs;
+use std::path::Path;
+use std::process::{Child, Command, Stdio};
+
+#[derive(Serialize, Debug)]
+struct LaunchVars<'a> {
+    user: &'a User,
+    node: i8,
+    current_time: String,
+}
+
+#[derive(Serialize, Debug)]
+struct BatchCommands {
+    commands: String,
+}
+
+fn get_door<'a>(name: &String, config: &'a Config) -> &'a Door {
+    let door = match config.doors.get(name) {
+        Some(door) => door,
+        None => {
+            eprintln!("Unknown door: {0}", name);
+            std::process::exit(1);
+        }
+    };
+
+    return door;
+}
+
+fn get_caller(config: &Config) -> User {
+    let user = match config.get_user(None) {
+        Ok(user) => user,
+        Err(e) => {
+            eprintln!("{0}", e);
+            std::process::exit(1);
+        }
+    };
+
+    return user;
+}
+
+fn switch_user(user: User, config: &Config, to_user: &Option<String>) -> User {
+    return match &to_user {
+        Some(username) => {
+            if !user.is_sysop {
+                eprintln!("You can't switch users if you're not a sysop!");
+                std::process::exit(1);
+            }
+            match config.get_user(Some(&username)) {
+                Ok(user) => user,
+                Err(e) => {
+                    eprintln!("{0}", e);
+                    std::process::exit(1);
+                }
+            }
+        }
+        None => user,
+    };
+}
+
+fn get_container_id(child: Child) -> String {
+    return match child.wait_with_output() {
+        Ok(output) => match output.status.code() {
+            Some(code) => {
+                if code != 0 {
+                    eprintln!("Starting container failed with status {0}", code);
+                    std::process::exit(1);
+                }
+                match String::from_utf8(output.stdout) {
+                    Ok(container_id) => container_id,
+                    Err(e) => {
+                        eprintln!("Couldn't decode container ID: {0}", e);
+                        std::process::exit(1);
+                    }
+                }
+            }
+            None => {
+                eprintln!("Starting container failed with unknown status");
+                std::process::exit(1);
+            }
+        },
+        Err(e) => {
+            eprintln!("Starting container failed: {0}", e);
+            std::process::exit(1);
+        }
+    };
+}
+
+pub fn launch(args: &LaunchArgs, config: &Config) {
+    let door = get_door(&args.door, config);
+    let user = switch_user(get_caller(config), config, &args.user);
+
+    let rundir = Path::new(&config.doorman.rundir);
```

Comment:
Instead of having to construct a path you can just have a `PathBuf` be in your CLI struct, I'm pretty sure.